### PR TITLE
KOGITO-8289: SWF Editor - Marshalling codebase clean-up

### PR DESCRIPTION
**Pull request by hasys**
Hello @romartin, @LuboTerifaj,

there is the refactoring of marshallers. I did a quick tests and didn't found any changes in kogito-showcase application examples. The marshaller logic is covered by Selenium tests, but if you think there is some missing Unit level tests, I will  add it to this PR. Thank you!

Task for this PR:

* Clean up unused classes
 * Move custom converters  [2] into the marshallers package [1]
 * Clean up states and remove unsupported parameters
 * Write Unit tests for logic where needed.

### CODE REVIEW

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/CallbackState.java [hunk 1]
```diff
@@ -35,14 +45,39 @@ public class CallbackState extends State {
...
+    public String compensatedBy;
+
     public CallbackState() {
         this.type = TYPE_CALLBACK;
     }
 
     public String getEventRef() {
         return eventRef;
     }
-
```
Comment (LuboTerifaj) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/CallbackState.java]: You probably removed this empty line by accident.

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/CallbackState.java [hunk 2]
```diff
@@ -16,16 +16,26 @@
...
+import org.kie.workbench.common.stunner.sw.marshall.yaml.StateTransitionDefinitionYamlTypeSerializer;
+import org.kie.workbench.common.stunner.sw.marshall.yaml.WorkflowTimeoutsYamlSerializer;
 
 @JSONMapper
 @YAMLMapper
 @JsType
-@YamlPropertyOrder({"name", "type", "transition", "action", "eventRef", "stateDataFilter", "eventTimeout", "compensatedBy", "timeouts", "onErrors", "end",  "metadata"})
-public class CallbackState extends State {
+@YamlPropertyOrder({"name", "type", "transition", "action", "eventRef", "stateDataFilter", "eventTimeout", "compensatedBy", "timeouts", "onErrors", "end", "metadata"})
+public class CallbackState extends State<CallbackState> implements HasEnd<CallbackState>, HasMetadata<CallbackState>, HasErrors<CallbackState>, HasCompensatedBy<CallbackState> {
```
Comment (LuboTerifaj) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/CallbackState.java]: Could you please apply KIE Java Conventions here?

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/EventState.java [hunk 3]
```diff
@@ -16,23 +16,59 @@
...
+import org.kie.workbench.common.stunner.sw.marshall.yaml.StateTransitionDefinitionYamlTypeSerializer;
+import org.kie.workbench.common.stunner.sw.marshall.yaml.WorkflowTimeoutsYamlSerializer;
 
 @JSONMapper
 @YAMLMapper
 @JsType
-@YamlPropertyOrder({"name", "type", "compensatedBy", "stateDataFilter", "timeouts", "end","exclusive", "onEvents", "onErrors", "eventTimeout", "transition", "metadata"})
-public class EventState extends State {
+@YamlPropertyOrder({"name", "type", "compensatedBy", "stateDataFilter", "timeouts", "end", "exclusive", "onEvents", "onErrors", "eventTimeout", "transition", "metadata"})
+public class EventState extends State<EventState> implements HasEnd<EventState>, HasMetadata<EventState>, HasErrors<EventState>, HasCompensatedBy<EventState> {
```
Comment (LuboTerifaj) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/EventState.java]: Could you please apply KIE Java Conventions here?

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/ForEachState.java [hunk 4]
```diff
@@ -24,14 +24,20 @@
...
+import org.kie.workbench.common.stunner.sw.marshall.yaml.StateTransitionDefinitionYamlTypeSerializer;
+import org.kie.workbench.common.stunner.sw.marshall.yaml.StringNumberYamlTypeSerializer;
+import org.kie.workbench.common.stunner.sw.marshall.yaml.WorkflowTimeoutsYamlSerializer;
 
 @JSONMapper
 @YAMLMapper
 @JsType
 @YamlPropertyOrder({"name", "type", "actions", "inputCollection", "outputCollection", "iterationParam", "transition", "stateDataFilter", "eventTimeout", "compensatedBy", "timeouts", "onErrors", "end", "metadata"})
-public class ForEachState extends State {
+public class ForEachState extends State<ForEachState> implements HasEnd<ForEachState>, HasMetadata<ForEachState>, HasErrors<ForEachState>, HasCompensatedBy<ForEachState> {
```
Comment (LuboTerifaj) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/ForEachState.java]: Could you please apply KIE Java Conventions here?

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/ForEachState.java [hunk 5]
```diff
@@ -51,14 +57,39 @@ public class ForEachState extends State {
...
+    public String compensatedBy;
+
     public String getMode() {
         return mode;
     }
 
     public void setMode(String mode) {
         this.mode = mode;
     }
-
```
Comment (LuboTerifaj) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/ForEachState.java]: This was removed probably by accident.

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/InjectState.java [hunk 6]
```diff
@@ -24,14 +24,18 @@
...
+import org.kie.workbench.common.stunner.sw.marshall.yaml.StateEndDefinitionYamlTypeSerializer;
+import org.kie.workbench.common.stunner.sw.marshall.yaml.StateTransitionDefinitionYamlTypeSerializer;
+import org.kie.workbench.common.stunner.sw.marshall.yaml.StringOrValueHolderYamlTypeSerializer;
 
 @JSONMapper
 @YAMLMapper
 @JsType
 @YamlPropertyOrder({"name", "type", "transition", "usedForCompensation", "stateDataFilter", "eventTimeout", "compensatedBy", "timeouts", "onErrors", "end", "data", "metadata"})
-public class InjectState extends State {
+public class InjectState extends State<InjectState> implements HasEnd<InjectState>, HasMetadata<InjectState>, HasCompensatedBy<InjectState> {
```
Comment (LuboTerifaj) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/InjectState.java]: Could you please apply KIE Java Conventions here?

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/OperationState.java [hunk 7]
```diff
@@ -16,16 +16,26 @@
...
+import org.kie.workbench.common.stunner.sw.marshall.yaml.StateEndDefinitionYamlTypeSerializer;
+import org.kie.workbench.common.stunner.sw.marshall.yaml.StateTransitionDefinitionYamlTypeSerializer;
+import org.kie.workbench.common.stunner.sw.marshall.yaml.WorkflowTimeoutsYamlSerializer;
 
 @JSONMapper
 @YAMLMapper
 @JsType
 @YamlPropertyOrder({"name", "type", "actions", "usedForCompensation", "transition", "end", "onErrors", "compensatedBy", "stateDataFilter", "timeouts", "eventTimeout", "actionMode", "metadata"})
-public class OperationState extends State {
+public class OperationState extends State<OperationState> implements HasEnd<OperationState>, HasMetadata<OperationState>, HasErrors<OperationState>, HasCompensatedBy<OperationState> {
```
Comment (LuboTerifaj) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/OperationState.java]: Could you please apply KIE Java Conventions here?

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/OperationState.java [hunk 8]
```diff
@@ -35,14 +45,39 @@ public class OperationState extends State {
...
+    public String compensatedBy;
+
     public OperationState() {
         this.type = TYPE_OPERATION;
     }
 
     public String getActionMode() {
         return actionMode;
     }
-
```
Comment (LuboTerifaj) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/OperationState.java]: This was probably removed by accident.

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/ParallelState.java [hunk 9]
```diff
@@ -24,14 +24,20 @@
...
+import org.kie.workbench.common.stunner.sw.marshall.yaml.StateTransitionDefinitionYamlTypeSerializer;
+import org.kie.workbench.common.stunner.sw.marshall.yaml.StringNumberYamlTypeSerializer;
+import org.kie.workbench.common.stunner.sw.marshall.yaml.WorkflowTimeoutsYamlSerializer;
 
 @JSONMapper
 @YAMLMapper
 @JsType
 @YamlPropertyOrder({"name", "type", "transition", "stateDataFilter", "compensatedBy", "branches", "timeouts", "eventTimeout", "onErrors", "end",  "metadata"})
-public class ParallelState extends State {
+public class ParallelState extends State<ParallelState> implements HasEnd<ParallelState>, HasMetadata<ParallelState>, HasErrors<ParallelState>, HasCompensatedBy<ParallelState> {
```
Comment (LuboTerifaj) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/ParallelState.java]: Could you please apply KIE Java Conventions here?

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/ParallelState.java [hunk 10]
```diff
@@ -49,14 +55,39 @@ public ParallelState() {
...
+    public String compensatedBy;
+
     public String getCompletionType() {
         return completionType;
     }
 
     public void setCompletionType(String completionType) {
         this.completionType = completionType;
     }
-
```
Comment (LuboTerifaj) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/ParallelState.java]: This was probably removed by accident.

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/SwitchState.java [hunk 11]
```diff
@@ -16,16 +16,22 @@
...
+import org.kie.workbench.common.stunner.client.yaml.mapper.api.annotation.YamlTypeSerializer;
+import org.kie.workbench.common.stunner.sw.marshall.json.WorkflowTimeoutsJsonSerializer;
+import org.kie.workbench.common.stunner.sw.marshall.yaml.WorkflowTimeoutsYamlSerializer;
 
 @JSONMapper
 @YAMLMapper
 @JsType
 @YamlPropertyOrder({"name", "type", "dataConditions", "eventConditions", "transition", "end", "onErrors", "usedForCompensation", "compensatedBy", "dataConditions", "stateDataFilter", "timeouts", "eventTimeout", "metadata"})
-public class SwitchState extends State {
+public class SwitchState extends State<SwitchState> implements HasMetadata<SwitchState>, HasErrors<SwitchState>, HasCompensatedBy<SwitchState> {
```
Comment (LuboTerifaj) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/SwitchState.java]: Could you please apply KIE Java Conventions here?

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/marshall/DocType.java [hunk 12]
```diff
@@ -17,5 +17,5 @@
 package org.kie.workbench.common.stunner.sw.marshall;
 
 public enum DocType {
-    JSON, XML, YAML;
+    JSON, XML, YAML
```
Comment (LuboTerifaj) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/marshall/DocType.java]:
```suggestion
    JSON,
    XML,
    YAML
```

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/marshall/Message.java [hunk 13]
```diff
@@ -35,9 +37,13 @@ public String getValue() {
...
     }
 
+    public String translateMessage(TranslationService translationService) {
+        return translationService.getTranslation(messageCode).replaceAll("%s", value);
+    }
+
     @Override
     public String toString() {
-        return messageCode.replaceAll("%s", value);
+        return "Message: {\"messageCode\": " +  messageCode + ", \"value\": " + value + "}";
```
Comment (LuboTerifaj) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/marshall/Message.java]:
```suggestion
        return "Message: {\"messageCode\": " + messageCode + ", \"value\": " + value + "}";
```

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/marshall/StateMarshalling.java [hunk 14]
```diff
@@ -82,39 +84,40 @@ public interface StateMarshalling {
...
                 }
 
-                // Parse on-errors.
-                ErrorTransition[] onErrors = state.getOnErrors();
-                if (null != onErrors && onErrors.length > 0) {
-                    for (int i = 0; i < onErrors.length; i++) {
-                        ErrorTransition onError = onErrors[i];
-                        if (null != onError) {
-                            Edge errorEdge = unmarshallEdge(context, onError);
+
```
Comment (LuboTerifaj) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/marshall/StateMarshalling.java]: Redundant empty line here.

**Review by LuboTerifaj — COMMENTED** [review]
Hello @hasys ,

I have just a few comments regarding code formatting.

Other than that, the changes look good to me.

Thank you!

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/marshall/StateMarshalling.java [hunk 15]
```diff
@@ -82,39 +84,40 @@ public interface StateMarshalling {
...
+                if (state instanceof HasErrors) {
```
Comment (romartin) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/marshall/StateMarshalling.java]: Hey @hasys 

I think by adding interfaces on the beans, they are harder to read and follow, at some point. The idea around the SWF domain modal has been always trying to make ti so so simple, without inheritance (as much as possible) and without implementing other types. In fact, I don't really like the idea of having getters and setters, as the model is finally transpiled to a structured javascript object, just with field declarations, no really need for getters and settters, as we do in java, it's just more dynamic code.

On the other hand I like the idea to introspect the beans in the marshaller operations, in order to figure out if has errors, etc. 

This way I propose the following:

- Drop the interfaces
- Refactor this kind  of lines, like `state instanceof HasErrors `, by just looking if the object contains the given field, like `errors`, then you can just set and get the field value, by using jsinterop, no need for getters/setters. Same thing you can do with `end`, `transition`, etc etc

IMO we will get the same results, although without all implementations and methods there, so just more clear... WDYT?

**Review by romartin — COMMENTED** [review]
Hey @hasys 

Code looks good!! Thanks.

Just added a single comment, although for me, important, please le me know what you think.

Thanks!

**Review by romartin — APPROVED** [review]
Great work @hasys , thanks for also adding the (test) stubs for Js stuff, looks nice!

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/marshall/StateMarshalling.java (hunk 15 above)
Comment (hasys) [reply packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/marshall/StateMarshalling.java]: Great point, fixed. Thank you1

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/marshall/StateMarshalling.java (hunk 14 above)
Comment (hasys) [reply packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/marshall/StateMarshalling.java]: Fixed, thank you!

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/marshall/Message.java (hunk 13 above)
Comment (hasys) [reply packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/marshall/Message.java]: Fixed, thank you!

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/marshall/DocType.java (hunk 12 above)
Comment (hasys) [reply packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/marshall/DocType.java]: Fixed, thank you!

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/SwitchState.java (hunk 11 above)
Comment (hasys) [reply packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/SwitchState.java]: Fixed, thank you!

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/ParallelState.java (hunk 10 above)
Comment (hasys) [reply packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/ParallelState.java]: Fixed, thank you!

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/InjectState.java (hunk 6 above)
Comment (hasys) [reply packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/InjectState.java]: Fixed, thank you!

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/OperationState.java (hunk 7 above)
Comment (hasys) [reply packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/OperationState.java]: Fixed, thank you!

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/OperationState.java (hunk 8 above)
Comment (hasys) [reply packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/OperationState.java]: Fixed, thank you!

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/ParallelState.java (hunk 9 above)
Comment (hasys) [reply packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/ParallelState.java]: Fixed, thank you!

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/ForEachState.java (hunk 5 above)
Comment (hasys) [reply packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/ForEachState.java]: Fixed, thank you!

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/ForEachState.java (hunk 4 above)
Comment (hasys) [reply packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/ForEachState.java]: Fixed, thank you!

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/EventState.java (hunk 3 above)
Comment (hasys) [reply packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/EventState.java]: Fixed, thank you!

**Review by LuboTerifaj — APPROVED** [review]
Hello @hasys , thank you for the updates. It looks good to me now.

**Comment by hasys** [comment]
Hello @kiegroup/gatekeepers PR is ready for merge. Thank you!

**Pull request merged**